### ISSUE TITLE
No way to obtain default apns-expiration behavior

### ISSUE THREAD

**Issue by brunchboy** [relayrides/pushy#566]
Pushy will [set](https://github.com/relayrides/pushy/blob/release-0.11.3/pushy/src/main/java/com/turo/pushy/apns/ApnsClientHandler.java#L233) an `apns-expiration` header with the value `0` when you send a notification with a `null` expiration timestamp. This tells Apple to not hold on to the notification at all, and give up immediately if the device is offline when it is sent. This is *different* than what Apple does if you simply omit the `apns-priority` header completely. In those circumstances, Apple uses a default expiration window, which is undocumented, but which empirical testing has shown to be at least a million seconds (Google has a similar default of 2419200 seconds); this behavior is discussed [here](https://stackoverflow.com/questions/44630196/what-is-the-default-value-of-the-apns-expiration-field).

It seems like it would be better for Pushy to simply not send this header at all when given a `null` expiration, so that clients can get the default behavior when they want it, and they can pass a `Date` that evaluates to a timestamp of 0 when they want an immediate expiration (and offer a class constant that contains such a date to make it easy).

**Comment by jchambers** [comment]
Interesting. Thanks for the research.

I definitely concede that what we're doing now means that callers can't really "ask for" the default behavior, but if nobody knows what the default behavior actually is, I'm not sure if that's much of a problem. I guess I'm thinking about what I'd write in the docs, and having a hard time thinking about how to frame it. I keep thinking it'd sound something like:

> …if `null`, then notifications will be stored by the APNs server for a default, but unknown and possibly-volatile period.

I wonder if we'd be better-served by introducing something like `ApnsPushNotification#DEFAULT_EXPIRATION_PERIOD`. At least then we could document what's happening with confidence.

What do you think?

**Comment by brunchboy** [comment]
Interesting, so you are proposing adding your own, documented period, and if someone supplies a `null` value for expiration, you create a `Date` representing that far into the future, and pass that to Apple. That would be easier to explain, understand, and rely on. I like it.

In the mean time, I am changing our code to use our own default interval so we never pass you `null`, as a quick fix.

Thanks, by the way, for this fantastic library!

**Comment by jchambers** [comment]
> Interesting, so you are proposing adding your own, documented period, and if someone supplies a null value for expiration, you create a Date representing that far into the future, and pass that to Apple.

Something along those lines, yeah. I think we'd also want to make sure that we have a clear way to express "don't store and forward" and consider backward compatibility, but that's the main idea.

Thanks for the feedback! Glad you're enjoying Pushy!

**Comment by brunchboy** [comment]
Yeah, I can see the backwards compatibility angle… it’s unfortunate that `null` can’t mean “Apple default” any longer because of that. Hmm… how about if the client can be configured with a default interval, and if you leave that unset, the behavior is as it is today, but if you set it to something then `null` means to send an expiration which is that far into the future?

Although that still leaves no way to request the Apple default behavior, but I guess I can concede your point that people should not rely on Apple’s default anyway since it is undocumented.

**Comment by jchambers** [comment]
I took a shot at this in #593. What do you think?
